packages/commonwealth/client/scripts/state/api/threads/helpers/cache.ts

```diff
@@ -0,0 +1,310 @@
+// import { useQueryClient } from '@tanstack/react-query';
+import Thread from 'models/Thread';
+import Topic from 'models/Topic';
+import { ApiEndpoints, queryClient } from 'state/api/config';
+
+type IExistingThreadState =
+  | null
+  | undefined
+  | { pages: any[]; pageParams: any[] }[]
+  | any;
+type IArrayManipulationMode =
+  | 'combineAndRemoveDups'
+  | 'removeFromExisting'
+  | 'replaceArray';
+
+interface CacheUpdater {
+  chainId: string;
+  threadId: number;
+  updateBody?: Partial<Thread>;
+  method: 'update' | 'remove';
+  arrayManipulationMode?: IArrayManipulationMode; // nested arrays are not updated
+}
+
+export const cacheTypes = {
+  SINGLE_THREAD: 'single',
+  BULK_THREADS: 'bulk',
+  ACTIVE_THREADS: 'active',
+};
+
+const cacheUpdater = ({
+  chainId,
+  threadId,
+  updateBody,
+  method,
+  arrayManipulationMode = 'replaceArray',
+}: CacheUpdater) => {
+  const queryCache = queryClient.getQueryCache();
+  const queryKeys = queryCache.getAll().map((cache) => cache.queryKey);
+  const keysForThreads = queryKeys.filter(
+    (x) => x[0] === ApiEndpoints.FETCH_THREADS && x[1] === chainId
+  );
+
+  const arrayFieldsFromUpdateBody = updateBody
+    ? Object.keys(updateBody).filter((k) => Array.isArray(updateBody[k]))
+    : [];
+
+  keysForThreads.map((k: any[]) => {
```

Comment:
As discussed, this piece of logic is effectively unreadable. Please clean it up and add documentation.

Reply:
I added some comments in this commit https://github.com/hicommonwealth/commonwealth/pull/4588/commits/79abe03fb86e4742b75eeed307be79561d2cb8f8. Please lmk if I should update more

Reply:
Good start, would like more comments within each of the methods, because that's also logic that's v difficult to follow. We may even want to pull them out into functions, so we can keep the logic a bit separated psychologically even if each of the functions is only called in one place.

Reply:
Good start, would like more comments within each of the methods, because that's also logic that's v difficult to follow. We may even want to pull them out into functions, so we can keep the logic a bit separated psychologically even if each of the functions is only called in one place.

Reply:
By works i mean https://github.com/hicommonwealth/commonwealth/issues/4664 is already updating threads API which will be reflected in frontend

Reply:
I don't mean utils, just named functions that live in the same file that separate out the various logical pieces. We should definitely add more comments regardless.

Reply:
hmm, i will move those, will need some time tho as I don't want to break anything there.

Reply:
it should be as easy as lifting the blocks of code into a couple functions -- no need to go crazy with it

Reply:
Ah, you meant that. That's done